docs/design/2022-06-15-phyical_plan_explain.md

```diff
@@ -0,0 +1,413 @@
+
+## Table of Contents
+
+- [Introduction](#introduction)
+- [Background](#background)
+- [Detailed Design](#detailed-design)
+- [Test Design](#test-design)
+
+## Introduction
+
+The physical plan is the physical execution plan in TiSpark. When we use explain in Spark which runs with TiSpark, the process of the physical plan will be displayed in the terminal. But now this display about how the plan executes has some problems.
+
+- `Residual Filter`  should not appear in physical plan explanation.
+- unclear representation of the execution process.
+- expression(s) that constitutes the scanning range is hard to know.
+- confusing naming of the operator.
+
+The display of the physical plan needs to be improved.
+
+## Background
+
+### The current way to build explanations
+
+If we call Spark's `explain`, we may see the following output.
+
+```text
+== Physical Plan ==
+*(1) ColumnarToRow
+// many other node.
++- ...
+	+- TiKV CoprocessorRDD{[table: t1] TableScan, Columns: a@UNSIGNED LONG, Residual Filter: [a@UNSIGNED LONG GREATER_THAN 1], PushDown Filter: [a@UNSIGNED LONG GREATER_THAN 1], KeyRange: [([t\200\000\000\000\000\000\004W_r\000\000\000\000\000\000\000\000], [t\200\000\000\000\000\000\004W_s\000\000\000\000\000\000\000\000])], Aggregates: , startTs: 433780573880188929}
+// many other node.
++- ...
+```
+
+What we are interested in here is the content in `CoprocessorRDD`. This node is the output of TiSpark, other nodes are the output of Spark. The content in the `CoprocessRDD` is the output of the `toStringInternal` method called by `TiDAGRequest`.
+
+First of all, let's introduce a few member variables in `TiDAGRequest`.
+
+* `indexInfo` is the index to be scanned by the physical plan, if it is empty, it means scanning the table.
+* `isDoubleRead` is `true` means scanning table after scanning the index, `false` means only scanning the index or scanning the table.
+* `filters` is the selection expression in the physical plan in normal execution.
+* `downgradeFilters` is the selection condition of the physical plan after the downgrade is triggered (what is downgrade will be described later).
+* `pushdownFilters` is only used in `toStringInternal`. When `indexInfo==null` or `isDoubleRead==false` or all the columns involved in the filter are in the index, the content of `pushDownFilters` is equal to filters, otherwise, `pushDownFilters` is empty.
+* `keyRange` is the scan range of the table or index.
+
+The main code is in [`toStringInternal`](https://github.com/pingcap/tispark/blob/v3.0.1/tikv-client/src/main/java/com/pingcap/tikv/meta/TiDAGRequest.java#L972-L1012).
+
+From the code we can see that the execution plan is divided into three kinds of `IndexScan`, `CoveringIndexScan`, and `TableScan`; `TiDAGRequest` has three kinds `Filter` —— `Downgrade Filter`, ` Resdiual Filter`, `PushDown Filter`. There is zero or more `Filter` for an execution plan. We will describe the meaning of each execution plan and `Filter` in detail below.
```

Comment:
From the code there is no `Resdiual Filter`. As you state below `Resdiual Filter` does not participate in `TiDAGRequest`. I think we should not show it here and just need to describe below.

Reply:
> From the code there is no `Resdiual Filter`. As you state below `Resdiual Filter` does not participate in `TiDAGRequest`. I think we should not show it here and just need to describe below.

The code has [`Resdiual Filter`](https://github.com/pingcap/tispark/blob/v3.0.1/tikv-client/src/main/java/com/pingcap/tikv/meta/TiDAGRequest.java#L972-L1012:~:text=if%20(!isIndexScan,%7D ).